Add tf.Model.summary()

caisq opened the pull request:
To this end, the following methods are also added:

- `get tf.Layer.outputShape`
- `tf.Layer.countParams`

FEATURE Add tf.Model.summary(): Print text summary of a tf.Model's topology and weight shapes.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-layers/238)
<!-- Reviewable:end -->

nsthorat commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

Nice work. This is super useful.

---

Review status: :shipit: complete! 1 of 1 LGTMs obtained

---

*[src/models.ts, line 537 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2w2qBD20Y4kq_h20L:-LF2w2qBD20Y4kq_h20M:b-rjic97) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/models.ts#L537)):*
> ```XML
>    *   `console.log`.
>    * @returns The printed summary text as an Array of strings.
>    */
> ```

add a @doc tag

---

*[src/models.ts, line 541 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2wIVeDK3jq4WYC0hA:-LF2wIVeDK3jq4WYC0hB:b-gkfu3h) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/models.ts#L541)):*
> ```XML
>       lineLength?: number, positions?: number[],
>       // tslint:disable-next-line:no-any
>       printFn: (message?: any, ...optionalParams: any[]) => void = console.log):
> ```

do we need this printFn if we return the string?

---

*[src/models.ts, line 542 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2wG1c6Jc23JoDGxw8:-LF2wG1c6Jc23JoDGxw9:b-o7wuwj) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/models.ts#L542)):*
> ```XML
>       // tslint:disable-next-line:no-any
>       printFn: (message?: any, ...optionalParams: any[]) => void = console.log):
>       string[] {
> ```

just curious, why string[] and not string with new lines?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-:-LF2wy6P4D2WjFmyVMlA:bw2heg6)*
<!-- Sent from Reviewable.io -->

bileschi commented:
Thanks!  Some thoughts on API and some smaller issues.

---







*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-:-LF2xqxDDJkJ70Ob-gEB:b-2k6nmp)*
<!-- Sent from Reviewable.io -->

bileschi commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r1, 3 of 3 files at r2.
Review status: :shipit: complete! 2 of 1 LGTMs obtained

---

*[src/model_summary_test.ts, line 11 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2cpbeAPK895qkysmd:-LF2cpbf9rbS64ztsW_w:b-ie0os5) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/model_summary_test.ts#L11)):*
> ```
> // Unit tests for tf.Model.summary() and tf.Sequential.summary().
> ```
To keep with convention, why not make this test and the utils file match names?  Rename this to "layer_utils_test.ts"?   Alternatively, rename "layer_utils.ts" to "model_summary.ts"?

---

*[src/model_summary_test.ts, line 45 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2vx77DNvkjcgit9rY:-LF2vx77DNvkjcgit9rZ:bg7ou6x) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/model_summary_test.ts#L45)):*
> ```
> ${layerName}
> ```
Two things.
1.  Kudos on making the default random layer name equal to the length of this variable plus syntax.  Very clever!
2.  I'm sure you're aware that these tests are brittle against any formatting changes in the summary.  Any thoughts about how a developer should update these if, e.g., spaces need to be turned to tabs or left-right justification needs to changed?

---

*[src/model_summary_test.ts, line 118 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2wk39DnNWc6itU0M6:-LF2wk39DnNWc6itU0M7:b-19z3ww) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/model_summary_test.ts#L118)):*
> ```
> 
> ```
add expect consoleLogHistory empty?

---

*[src/model_summary_test.ts, line 154 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2xUqx7k8yW0FyDqHR:-LF2xUqx7k8yW0FyDqHS:b-bixm4s) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/model_summary_test.ts#L154)):*
> ```
> lyrName01
> ```
lol gotta keep those 9 character variable names ;)

---

*[src/models.ts, line 401 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2TWOe1m3TKb7bJZUR:-LF2TWOe1m3TKb7bJZUS:b7dcjxn) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/models.ts#L401)):*
> ```
>  this.inboundNodes = [];
> ```
Not sure what's going on here?  

---

*[src/engine/topology.ts, line 1006 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2LhoN1Fyt4IzDxABz:-LF2LhoN1Fyt4IzDxAC-:b-v0lk12) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/engine/topology.ts#L1006)):*
> ```
> getOutputShapeAt().
> ```
More info than just the fn name needed here

---

*[src/engine/topology_test.ts, line 710 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2Mvaz6PWB4RaF3lGH:-LF2Mvaz6PWB4RaF3lGI:ba0oog6) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/engine/topology_test.ts#L710)):*
> ```
> it('Layers with two inboundNodes of the same outputShape'
> ```
Missing test to capture error behavior of two different output shapes

---

*[src/engine/training.ts, line 679 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2pX4S8OVytsxiobPh:-LF2pX4S8OVytsxiobPi:bep7kfl) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/engine/training.ts#L679)):*
> ```
> Custom line length.
> ```
in terminal characters

---

*[src/engine/training.ts, line 680 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2p_FeDsMnGJrrPTdx:-LF2p_FeDsMnGJrrPTdy:b57leb3) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/engine/training.ts#L680)):*
> ```
>  Custom widths of each of the columns.
> ```
May be fractional or absolute

---

*[src/engine/training.ts, line 681 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2peam6Ng3fHB2MOnK:-LF2peam6Ng3fHB2MOnL:b-ms6uzj) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/engine/training.ts#L681)):*
> ```
> replac
> ```
replace.  

Also, maybe add something here about how to get the summary return value without printing to console, something like (x) => {};

---

*[src/utils/generic_utils.ts, line 454 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2Q_Gn8-2F8pLzUzKz:-LF2Q_Gn8-2F8pLzUzL-:baozo6z) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/generic_utils.ts#L454)):*
> ```
> weight.shape
> ```
Are there circumstances in which a LayerVariable's shape could report negative elements?  Playing around with it it seems unlikely.  Just asking.

---

*[src/utils/generic_utils_test.ts, line 298 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2Rajx6Mw_ds7jJ4Bs:-LF2Rajx6Mw_ds7jJ4Bt:b-rz426d) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/generic_utils_test.ts#L298)):*
> ```
> Ont
> ```
nit: One

---

*[src/utils/layer_utils.ts, line 19 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2dx6j4FaA9RPSdegA:-LF2dx6j4FaA9RPSdegB:b-3dsepw) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L19)):*
> ```XML
>  *
>  * @param model tf.Model instance.
>  * @param lineLength Total length of printed lines. Set this to adapt to the
> ```

length in columns?  How are multi-column characters handled?  (It's ok if we don't handle them well, right away, but we should mention it)  https://unix.stackexchange.com/questions/245013/get-the-display-width-of-a-string-of-characters

---

*[src/utils/layer_utils.ts, line 22 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2dOrs1qFf6ak53ps1:-LF2dOrs1qFf6ak53ps2:b-b07fw) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L22)):*
> ```XML
>  *   display to different terminal or console sizes.
>  * @param positions Relative or absolute positions of log eleemnts in each
>  *   line.
> ```

the description of the `positions` argument is confusing.  What do the boundaries mean?  The edges of columns?  How wide each section is?  Why the trailing 1 but no preceding 0?

  Also (nit) "elements" has a typo.

---

*[src/utils/layer_utils.ts, line 48 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2iJxB-yeodtlm6o8f:-LF2iJxB-yeodtlm6o8g:b159u59) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L48)):*
> ```XML
>     // Header names for different log elements.
>   }
>   if (positions[positions.length - 1] <= 1) {
> ```

comment here:
// If positions is relative, convert to absolute positioning.

---

*[src/utils/layer_utils.ts, line 54 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2goHf4Haxkp5F5ZgZ:-LF2goHf4Haxkp5F5Zg_:b-6fc3ni) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L54)):*
> ```
> Recevies
> ```
nit: spelling.

---

*[src/utils/layer_utils.ts, line 87 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2kVjq7fr3Sb7PJJW4:-LF2kVjq7fr3Sb7PJJW5:b7gxihy) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L87)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
>     trainableCount =
>         countParamsInWeights((model as any).collectedTrainableWeights);
>   } else {
>     trainableCount = countParamsInWeights(model.trainableWeights);
> ```
</details>

Not sure I understand what is going on here?   Is this managing some sort of dispatch, determining which `collectedTrainableWeights` gets called?

---

*[src/utils/layer_utils.ts, line 155 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2lc2p8-qpPjRVIkZj:-LF2lc2p8-qpPjRVIkZk:bihb4du) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L155)):*
> ```XML
>   }
>   printFn(line);
>   return line;
> ```

Seems weird to me that the printing of the summary contents and the construction of the summary contents are intertwined.  I wonder why not construct all the lines and then print them at the end (if one wished)?

I guess you can get around this by selecting a printFn that's a noop, but thats clearly extra load if you just want the string value of the summary.

---

*[src/utils/layer_utils.ts, line 171 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2mYbI-17qCl73MhkB:-LF2mYbI-17qCl73MhkC:b-kam06) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L171)):*
> ```
>  outputShape = 'multiple';
> ```
?  What does 'multiple mean in this context?

---

*[src/utils/layer_utils.ts, line 174 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2mcpL2N6yd5PPgf0q:-LF2mcpL2N6yd5PPgf0r:b35thrw) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L174)):*
> ```
>  const name = layer.name;
> ```
layer.name is optional, may want a pretty place-holder for nameless layers.

---

*[src/utils/layer_utils.ts, line 209 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2nF_H9UAz-YdCQrwo:-LF2nF_H9UAz-YdCQrwp:b-qn6b35) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L209)):*
> ```
> const name = layer.name;
> ```
name is optional, may want a placeholder

---

*[src/utils/layer_utils.ts, line 218 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2p0g9-ISCqsJHUvfg:-LF2p0g9-ISCqsJHUvfh:b-ooc0ny) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L218)):*
> ```
> if (connections.length > 1) {
> ```
not technically needed.  will be captured by predicate of for statement.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-:-LF2xvhrETvTNQ0jnoHn:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: :shipit: complete! 2 of 1 LGTMs obtained

---

*[src/model_summary_test.ts, line 11 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2cpbeAPK895qkysmd:-LF3SZAl8zLwiuJQNNQd:b9u61pc) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/model_summary_test.ts#L11)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> // Unit tests for tf.Model.summary() and tf.Sequential.summary().
> ```
To keep with convention, why not make this test and the utils file match names?  Rename this to "layer_utils_test.ts"?   Alternatively, rename "layer_utils.ts" to "model_summary.ts"?
</blockquote></details>

IMO, layer_utils.ts is not a very informative name. But I kept it for consistency with Keras. I separate the Model.summary() unit tests so that models_test.ts won't get too large. When separating the unit tests into a new file, I took the liberty to name it to something more appropriate and informative, namely model_summary_test.ts. The file name also has symmetry with an existing test file, namely model_save_test.ts.

---

*[src/model_summary_test.ts, line 45 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2vx77DNvkjcgit9rY:-LF3TEKm3n09BDAmhsyE:b-7oz3nw) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/model_summary_test.ts#L45)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> ${layerName}
> ```
Two things.
1.  Kudos on making the default random layer name equal to the length of this variable plus syntax.  Very clever!
2.  I'm sure you're aware that these tests are brittle against any formatting changes in the summary.  Any thoughts about how a developer should update these if, e.g., spaces need to be turned to tabs or left-right justification needs to changed?
</blockquote></details>

Hopefully developers won't have to update this too often. The formatting is more or less stable, thanks to Keras. 
In case someone needs to update it, he/she can always manually update these values, possibly by copy-pasting from a terminal (after manual examination to make sure everything is correct, of course.)

---

*[src/model_summary_test.ts, line 118 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2wk39DnNWc6itU0M6:-LF3Teps5ntIBudbN2cN:boj1qgt) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/model_summary_test.ts#L118)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> 
> ```
add expect consoleLogHistory empty?
</blockquote></details>

Good catch. Done.

---

*[src/model_summary_test.ts, line 154 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2xUqx7k8yW0FyDqHR:-LF3TogM8Mt59qZYBjP0:b-rcfd5n) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/model_summary_test.ts#L154)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> lyrName01
> ```
lol gotta keep those 9 character variable names ;)
</blockquote></details>

Ack. :)

---

*[src/models.ts, line 401 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2TWOe1m3TKb7bJZUR:-LF3UycL55BCPaILdHoO:b-x3z6lj) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/models.ts#L401)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
>  this.inboundNodes = [];
> ```
Not sure what's going on here?  
</blockquote></details>

This is a minor bug fix. When a layer is added to an empty sequential layer, the inboundNodes field of the sequential should be emptied first. Previously, it contained an extraneous item, causing issues during summary printing.

---

*[src/models.ts, line 537 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2w2qBD20Y4kq_h20L:-LF3VM6X4agJr5O8s9RA:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/models.ts#L537)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

add a @doc tag
</blockquote></details>

Done.

---

*[src/models.ts, line 541 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2wIVeDK3jq4WYC0hA:-LF3Vb5W2HMQECfuU3wo:b-du5h5y) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/models.ts#L541)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

do we need this printFn if we return the string?
</blockquote></details>

See my reply to Stan below as well. Good point. I think there is some redundancy in `printFn` and the string return value. Redundancy in API is generally not a good thing, because it is confusing. I removed the string return value.

---

*[src/models.ts, line 542 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2wG1c6Jc23JoDGxw8:-LF3W3BGC4Ki-ivWQfSg:b-twkho3) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/models.ts#L542)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

just curious, why string[] and not string with new lines?
</blockquote></details>

The return value is removed. See my reply to the question above.

---

*[src/engine/topology_test.ts, line 710 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2Mvaz6PWB4RaF3lGH:-LF3XZC69RmWgV0YKIjJ:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/engine/topology_test.ts#L710)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> it('Layers with two inboundNodes of the same outputShape'
> ```
Missing test to capture error behavior of two different output shapes
</blockquote></details>

Done.

---

*[src/engine/training.ts, line 679 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2pX4S8OVytsxiobPh:-LF3XzjDAzS91pL1OWpq:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/engine/training.ts#L679)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> Custom line length.
> ```
in terminal characters
</blockquote></details>

Done.

---

*[src/engine/training.ts, line 680 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2p_FeDsMnGJrrPTdx:-LF3XzClAxgEiXs2nJZ7:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/engine/training.ts#L680)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
>  Custom widths of each of the columns.
> ```
May be fractional or absolute
</blockquote></details>

Done.

---

*[src/engine/training.ts, line 681 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2peam6Ng3fHB2MOnK:-LF3YJazCyKNZDe1ATnO:b-qu6n94) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/engine/training.ts#L681)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> replac
> ```
replace.  

Also, maybe add something here about how to get the summary return value without printing to console, something like (x) => {};
</blockquote></details>

Good point. Done.

---

*[src/utils/generic_utils.ts, line 454 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2Q_Gn8-2F8pLzUzKz:-LF3YULFCYDh_CBDAN_N:bom5g82) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/generic_utils.ts#L454)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> weight.shape
> ```
Are there circumstances in which a LayerVariable's shape could report negative elements?  Playing around with it it seems unlikely.  Just asking.

</blockquote></details>

A LayerVariable holds a concrete tensor value. So all of its dimensions should be positive integers.

---

*[src/utils/generic_utils_test.ts, line 298 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2Rajx6Mw_ds7jJ4Bs:-LF3YdzA3zyQG9h3GZlR:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/generic_utils_test.ts#L298)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> Ont
> ```
nit: One
</blockquote></details>

Done.

---

*[src/utils/layer_utils.ts, line 19 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2dx6j4FaA9RPSdegA:-LF3aYpS9rAlAuJwhYZS:b-ph7xoc) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L19)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

length in columns?  How are multi-column characters handled?  (It's ok if we don't handle them well, right away, but we should mention it)  https://unix.stackexchange.com/questions/245013/get-the-display-width-of-a-string-of-characters
</blockquote></details>

You cannot use wide characters in tensor names currently: 
https://github.com/tensorflow/tfjs-layers/blob/master/src/common.ts#L115

---

*[src/utils/layer_utils.ts, line 22 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2dOrs1qFf6ak53ps1:-LF3a29Q-67NaKr3Tqr4:bd6cnz1) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L22)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

the description of the `positions` argument is confusing.  What do the boundaries mean?  The edges of columns?  How wide each section is?  Why the trailing 1 but no preceding 0?

  Also (nit) "elements" has a typo.
</blockquote></details>

Added "Each number corresponds to right-most (i.e., ending) position of a column."
Also added to the documentation of Model.summary() and Sequential.summary().
Typo corrected.

---

*[src/utils/layer_utils.ts, line 48 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2iJxB-yeodtlm6o8f:-LF3bABqElJMCSUL1xPo:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L48)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

comment here:
// If positions is relative, convert to absolute positioning.
</blockquote></details>

Done.

---

*[src/utils/layer_utils.ts, line 54 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2goHf4Haxkp5F5ZgZ:-LF3aRcx2svXENJoXIwd:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L54)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> Recevies
> ```
nit: spelling.
</blockquote></details>

Done.

---

*[src/utils/layer_utils.ts, line 87 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2kVjq7fr3Sb7PJJW4:-LF3chLGAX8w7hBHrJQq:b-9ilvuu) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L87)):*
> ```
> collectedTrainableWeights
> ```

collectedTrainableWeights is populated during Model.compile() calls. It is a flat array of weights and is more efficient to retrieve, compared to calling model.trainableWeights, which needs to recursively go into all the constituent layers. This code here is just getting the trainable weights via the most efficient route, depending on whether collectedTrainableWeights is available.

---

*[src/utils/layer_utils.ts, line 155 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2lc2p8-qpPjRVIkZj:-LF3dp2O0vHqO9rUSyYE:b-khjzbl) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L155)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

Seems weird to me that the printing of the summary contents and the construction of the summary contents are intertwined.  I wonder why not construct all the lines and then print them at the end (if one wished)?

I guess you can get around this by selecting a printFn that's a noop, but thats clearly extra load if you just want the string value of the summary.

</blockquote></details>

Good point. I think there is some redundancy in `printFn` and the string return value. Redundancy in API is generally not a good thing, because it is confusing. I removed the string return value.

---

*[src/utils/layer_utils.ts, line 171 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2mYbI-17qCl73MhkB:-LF3bPw7-g6edbZRWcqo:bn8wcrz) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L171)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
>  outputShape = 'multiple';
> ```
?  What does 'multiple mean in this context?
</blockquote></details>

It means there are multiple output shapes due to multiple inbound nodes to the layer.

---

*[src/utils/layer_utils.ts, line 174 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2mcpL2N6yd5PPgf0q:-LF3b_L98Z3F052OoRFo:b4yx2o8) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L174)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
>  const name = layer.name;
> ```
layer.name is optional, may want a pretty place-holder for nameless layers.
</blockquote></details>

name is optional in Layer's constructor. But after it is constructed, the name is always available.

---

*[src/utils/layer_utils.ts, line 209 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2nF_H9UAz-YdCQrwo:-LF3beUp6ARlPnNXnR8b:bc096wy) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L209)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> const name = layer.name;
> ```
name is optional, may want a placeholder
</blockquote></details>

name is optional in Layer's constructor. But after it is constructed, the name is always available.

---

*[src/utils/layer_utils.ts, line 218 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2p0g9-ISCqsJHUvfg:-LF3bnmw4qLemxNEVMrq:b-llqefw) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L218)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> if (connections.length > 1) {
> ```
not technically needed.  will be captured by predicate of for statement.
</blockquote></details>

That's true. Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/238)*
<!-- Sent from Reviewable.io -->

bileschi commented:
*[src/utils/layer_utils.ts, line 155 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2lc2p8-qpPjRVIkZj:-LF3p68bDQHQTApBLEAY:b-yqtmgw) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L155)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

Good point. I think there is some redundancy in `printFn` and the string return value. Redundancy in API is generally not a good thing, because it is confusing. I removed the string return value.
</blockquote></details>

The signature was removed, but the value is still returned?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/238)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Review status: :shipit: complete! 2 of 1 LGTMs obtained

---

*[src/models.ts, line 542 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2wG1c6Jc23JoDGxw8:-LF3qDr4AWYgeCyoR6Lv:bb2cq4f) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/models.ts#L542)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

The return value is removed. See my reply to the question above.
</blockquote></details>

Any reason why to prefer the callback over the simple return type? I think the return type is simpler. WDYT?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/238)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: :shipit: complete! 2 of 1 LGTMs obtained

---

*[src/models.ts, line 542 at r2](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2wG1c6Jc23JoDGxw8:-LF4GerBCwhduBFQ0_fA:b6npibp) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/f0d854232156c3d76df13afc28c92663f2db29e9/src/models.ts#L542)):*
> simpler

If we are to keep only one of the return value and the custom printFn, I prefer the latter, because 1) this is consistent with Keras and 2) it is more flexible (imagine a case in which you don't want to get hold of the string values, but just want to write them to a file).

---

*[src/utils/layer_utils.ts, line 155 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/238#-LF2lc2p8-qpPjRVIkZj:-LF4GVlt8nSS2HvNm_Os:bm7fvnk) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/d19426f86b0a2c890a5c35824ac7ece005b3bba7/src/utils/layer_utils.ts#L155)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

The signature was removed, but the value is still returned?
</blockquote></details>

Good catch. Removed the `return`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/238)*
<!-- Sent from Reviewable.io -->
